pymc_marketing/mmm/lift_test.py

```diff
@@ -0,0 +1,366 @@
+from typing import Callable, Optional
+
+import numpy as np
+import pandas as pd
+import pymc as pm
+import pytensor.tensor as pt
+from numpy import typing as npt
+
+from pymc_marketing.mmm.transformers import logistic_saturation
+from pymc_marketing.mmm.utils import michaelis_menten
+
+
+class MissingLiftTestError(Exception):
+    def __init__(self, missing_values: npt.NDArray[np.int_]) -> None:
+        self.missing_values = missing_values
+        super().__init__(
+            f"Some lift test values are not in the model: {missing_values}"
+        )
+
+
+def _lift_test_index(
+    lift_values: np.ndarray, model_values: np.ndarray
+) -> npt.NDArray[np.int_]:
+    same_value = lift_values[:, None] == model_values
+    if not (same_value.sum(axis=1) == 1).all():
+        missing_values = np.argwhere(same_value.sum(axis=1) == 0).flatten()
+        raise MissingLiftTestError(missing_values)
+
+    return np.argmax(same_value, axis=1)
+
+
+Index = npt.NDArray[np.int_]
+Indices = dict[str, Index]
+
+
+def lift_test_indices(df_lift_test: pd.DataFrame, model: pm.Model) -> Indices:
+    """Get the indices of the lift test results in the model.
+
+    Assumes any column in the DataFrame is a coordinate in the model with the
+    same name.
+
+    Args:
+        df_lift_test: DataFrame with lift test results.
+        model: PyMC model with date and channel coordinates.
+
+    Returns:
+        Dictionary of indices for the lift test results in the model.
+
+    """
+    columns = df_lift_test.columns.tolist()
+    return {
+        col: _lift_test_index(df_lift_test[col].to_numpy(), np.array(model.coords[col]))
+        for col in columns
+    }
+
+
+def empirical_lift_measurements(x_before, x_after, saturation_curve, pt=pt):
+    """Calculate the empirical lift measurements at two spends.
+
+    Args:
+        x_before: Array of x before the change.
+        x_after: Array of x after the change.
+        saturation_curve: Function that takes spend and returns saturation.
+        pt: PyTensor module.
+
+    Returns:
+        Array of empirical lift measurements.
+
+    """
+    return pt.diff(
+        saturation_curve(pt.stack([x_before, x_after])),
+        axis=0,
+    ).flatten()
+
+
+def required_dims_from_named_vars_to_dims(
+    named_vars_to_dims: dict[str, tuple[str, ...]],
+) -> list[str]:
+    """Get the required dimensions from a named_vars_to_dims dictionary.
+
+    Args:
+        named_vars_to_dims: Dictionary of variable names to dimensions.
+
+    Returns:
+        List of required dimensions.
+
+    """
+    required_dims = set()
+    for dims in named_vars_to_dims.values():
+        for dim in dims:
+            required_dims.add(dim)
+
+    return list(required_dims)
+
+
+def indices_from_lift_tests(
+    df_lift_test: pd.DataFrame,
+    model: pm.Model,
+    var_names: list[str],
+) -> Indices:
+    """Get the indices of the lift test results in the model.
+
+    Args:
+        df_lift_test: DataFrame with lift test results.
```

Comment:
It would be good to provide the desired structure of the dataframe: size? columns? types?